packages/swingset-liveslots/src/boyd-gc.js

```diff
@@ -27,138 +28,393 @@ import { parseVatSlot } from './parseVatSlots.js';
+  three pillars:
+  * the RAM pillar is the `Representative` object
+  * the vdata pillar is stored in `vom.rc.${vref}`
+  * the export-status pillar is stored in `vom.es.${vref}`
+
+  Remotables have only the RAM pillar. When a Remotable-style vref is
+  exported to the kernel, the Remotable is added to the
+  `exportedRemotables` set. And when it is stored in vdata, it appears
+  as a key of the `remotableRefCounts` map. That keeps the Remotable
+  itself alive until the other reachability pathways have gone
+  away. We don't do this for Representatives because it would violate
+  our "don't use RAM for inactive virtual objects" rule.
+
+  When an imported Presence becomes UNREACHABLE, it might still be
+  RECOGNIZABLE, by virtue of being the key of one or more weak
+  collections. If not, it might transit from REACHABLE directly to
+  NONE. The code that reacts to the UNREACHABLE transition must check
+  for recognizers, and do a retireImports right away if there are
+  none. Otherwise, recognizer will remain until either the kernel
+  retires the object (dispatch.retireImports), or the weak collection
+  is deleted, in which case possiblyRetiredSet will be updated with
+  the vref that might no longer be recognized. There will be a race
+  between us ceasing to recognize the vref (which should trigger a
+  syscall.retireImports), and the kernel telling us the object has
+  been deleted (via dispatch.retireImports). Either one must inhibit
+  the other.
+
+  possiblyRetiredSet only cares about Presence-style vrefs, because
+  they represent imports, whose lifetime is not under our control. The
+  collection-deletion code will add Remotable- and Representative-
+  style vrefs in possiblyRetiredSet, but we can remove and ignore
+  them.
+
+  We use slotToVal.has(vref) everywhere for our "is it still
+  reachable" check, which returns true for the REACHABLE / UNREACHABLE
+  / COLLECTED states, and false for the FINALIZED state. In contrast,
+  getValForSlot(vref) returns false for both COLLECTED and
+  FINALIZED. We want COLLECTED to qualify as "still reachable" because
+  it means there's still a finalizer callback queued, which will be
+  run eventually, and we need that callback to not trigger a duplicate
+  drop. We use slotToVal.has() in the possiblyRetiredSet loop (to
+  inhibit retirement of imported vrefs whose Presences are in the
+  COLLECTED state, and which just lost a recognizer), because
+  getValForSlot would allow such COLLECTED vrefs to be retired, even
+  before the finalizer had fired and could do a dropImports.
 
+  Deleting a virtual object will delete its `state`, decrementing the
+  refcounts of any objects therein, which might shake loose more
+  data. So we keep looping until we've stopped adding things to
+  possiblyDeadSet. Retiring a Presence-style vref, on the other hand,
+  cannot shake anthing loose. So we do not include the
+  possiblyRetiredSet processing inside that loop.
```

Comment:
The `possiblyRetiredSet` loop's only job is to perform `syscall.retireImports` on the imported vrefs that have transitioned from "the vat cannot reach this, but might still recognize it" to "the vat cannot even recognize it". The issuance of a `syscall.retireImports` cannot shake anything loose in the vat.

It is the previous loop, the one that walks `possiblyDeadSet`, where the shaking-loose phase happens. When that loop examines a Presence-style import vref (o-1), in the `checkExportRemotable`helper function, if it is found to be unreachable, it calls `vrm.ceaseRecognition`. That function locates all the recognition records (to identify all the weak collections which are using the vref as a key), and tells the collections to delete that entry (with either `recognizer.delete(vref)` for the voAwareWeakMap/Set ones, or `deleteCollectionEntry()` for the virtual/durable weak stores). Those deletion functions will decrement the value's refcounts, just as if someone did `collection.delete(presence)` (which, of course, the can no longer do: we're only here because `presence` has gone out of scope).

The function and variable names aren't ideal. I frequently wonder whether using "retire" to talk about both imported (Presence-style) vrefs and locally-decided (Remotable/Representative -style) vrefs is a good idea. `dispatch.retireImports` seems right ("hi vat, the object you know as vref o-4 has retired, stop thinking about it"). And `syscall.dropImports` seems right ("hi kernel, I'm no longer interested in keeping the object I know of as vref o-2 alive, because nobody here can reach it anymore, feel free to, you know, tell someone about it"). But `syscall.retireImports` is starting to sound weird, it should be more like "cease recognition" or `unrecognizeImports` ot `derecognizeImports`. But the fact that each vref should appear in either `dispatch.retireImports` or `syscall.retireImports`, never both, suggests that they should share a name.

In response to one of @gibson042 's comments on the second PR, I wrote two new pages of docs for this file, and started using the phrase "logical object" to describe the thing that is 1: associated with a vref, 2: supported by pillars, 3: survives the original JS object going away, and 4: transcends vat boundaries. That makes it easier to say "the logical object is reachable, but the Presence is COLLECTED/FINALIZED", but I had to start using distinct state names like VREF-REACHABLE to distinguish from PRESENCE-REACHABLE or EXPORT-REACHABLE.

Anyways, this comment could probably be better. If we read "Retiring a Presence-style vref" as "the logical object with an o-NN vref has been retired, now react to that", then yeah, we delete weak collection entries when the key is retired, so it would imply we're going to do that `ceaseRecognition`, which will indeed shake things loose. What I really meant in this section was "we've stopped recognizing this o-NN vref, now we just need to tell the kernel about it".

Lemme change that sentence, and make it clear that we're reacting to things that have been retired, and that the deletions and decrefs have already been done, so the remaining responsibility is to notify the kernel about the now-unrecognizable vrefs.

Does that address your concern?

Reply:
Oh, the comments are missing something important. When a Presence-style logical object is deleted, it's the kernel (really the upstream/exporting vat) which makes that decision. From down here at the importing side, we merely get *notified* by the kernel that the vref has been retired, with a `dispatch.retireImports`. *That* function, over in liveslots.js, has a call to `vrm.ceaseRecognition()`, which does the deletion of weak-collection values.

Here, in BOYD, we don't get to decide when a `o-NN` logical object gets deleted. We merely decide when this vat cannot reach it anymore, and react to that transition (VREF-REACHABLE to VREF-RECOGNIZABLE, in today's latest jargon) by telling the kernel about it, with a `syscall.dropImports`.

It's weird, because for Representative-style vrefs, we call `ceaseRecognition` here, in BOYD, but for Presence-style ones, the same function gets called from a `dispatch` handler. They're different because the kernel decides (and deigns to inform us) about Presence-style logical objects being deleted, whereas for Representative-style, we must make that decision. And to make it correctly, we need to know the finalizer state, making us sensitive to GC behavior, in a way that we're only willing to do during a BOYD. So the syscalls that are sensitive to GC (all the vatstoreDelete and refcount changes) must happen during BOYD, and thus inside scanForDeadObjects, and they really happen inside checkExportRepresentative.

I'll see if I can capture that in the docs too.

Reply:
new text, which no longer mentions "we do not need to include the possiblyRetiredSet processing inside that loop":


  When we decide to delete a virtual object, we will delete its
  `state`, decrementing the refcounts of any objects therein, which
  might shake loose more data. So we keep looping until we've stopped
  adding things to possiblyDeadSet. The same can happen when we use
  vrm.ceaseRecognition() to delete any weak collection values keyed by
  it. We also call ceaseRecognition when we realize that a Remotable
  has been deleted. But the possiblyDeadSet processing loop cannot
  make the decision to retire a Presence-style vref: those are deleted
  outside our vat, and the kernel notifies us of the vref's retirement
  with dispatch.retireImports (which also calls ceaseRecognition). The
  only thing possiblyDeadSet can tell us about Presences is that our
  vat can no longer *reach* the vref, which means we need to do a
  syscall.dropImports, which cannot immediately release more data.

Reply:
perfect. And tracing retention paths as described in #6793 is exactly what this would solve: avoiding the multi-step GC and instead let the kernel tell the vat "you are safe to retire this set of objects that you previously couldn't reach by yourself".